bigquery/bigquery_analyticshub_quickstart/main.go

```diff
@@ -0,0 +1,167 @@
+
+	dataexchange "cloud.google.com/go/bigquery/dataexchange/apiv1beta1"
+	"google.golang.org/genproto/googleapis/cloud/bigquery/dataexchange/common"
+	dataexchangepb "google.golang.org/genproto/googleapis/cloud/bigquery/dataexchange/v1beta1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func main() {
+
+	// Define two command line flags for controlling the behavior of this quickstart.
+	var (
+		projectID            = flag.String("project_id", "", "Cloud Project ID, used for session creation.")
+		location             = flag.String("location", "US", "BigQuery location used for interactions.")
+		exampleExchangeID    = flag.String("exchange_id", "ExampleDataExchange", "identifier of the example data exchange")
+		exampleListingID     = flag.String("listing_id", "ExampleDataExchange", "identifier of the example data exchange")
+		exampleDatasetSource = flag.String("dataset_source", "", "dataset source in the form projects/myproject/datasets/mydataset")
+		delete               = flag.Bool("delete_exchange", true, "delete exchange at the end of quickstart")
+	)
+	flag.Parse()
+	// Perform simple validation of the specified flags.
+	if *projectID == "" {
+		log.Fatal("empty --project_id specified, please provide a valid project ID")
+	}
+	if *exampleDatasetSource == "" {
+		log.Fatalf("empty --dataset_source specified, please provide in the form \"projects/myproject/datasets/mydataset\"")
+	}
+
+	// Instantiate the client.
+	ctx := context.Background()
+	dataExchClient, err := dataexchange.NewAnalyticsHubClient(ctx)
+	if err != nil {
+		log.Fatalf("NewClient: %v", err)
+	}
+	defer dataExchClient.Close()
+
+	// Then, create the data exchange (or return information about one already bearing the example name), and
+	// print information about it.
+	exchange, err := createOrGetDataExchange(ctx, dataExchClient, *projectID, *location, *exampleExchangeID)
+	if err != nil {
+		log.Fatalf("failed to get information about the exchange: %v", err)
+	}
+	fmt.Printf("\nData Exchange Information\n")
+	fmt.Printf("Exchange Name: %s\n", exchange.GetName())
+	if desc := exchange.GetDescription(); desc != "" {
+		fmt.Printf("Exchange Description: %s", desc)
+	}
+
+	// Finally, create a listing within the data exchange and print information about it.
+	listing, err := createListing(ctx, dataExchClient, *projectID, *location, *exampleExchangeID, *exampleListingID, *exampleDatasetSource)
+	if err != nil {
+		log.Fatalf("failed to create the listing within the exchange: %v", err)
+	}
+	fmt.Printf("\n\nListing Information\n")
+	fmt.Printf("Listing Name: %s\n", listing.GetName())
+	if desc := listing.GetDescription(); desc != "" {
+		fmt.Printf("Listing Description: %s\n", desc)
+	}
+	fmt.Printf("Listing State: %s\n", listing.GetState().String())
+	if source := listing.GetSource(); source != nil {
+		if dsSource, ok := source.(*dataexchangepb.Listing_BigqueryDataset); ok {
+			if dataset := dsSource.BigqueryDataset.GetDataset(); dataset != "" {
```

Comment:
Can BigqueryDataset be nil?

Reply:
It's a oneof, so in theory there may at some point be another kind of shared resource in the listing.

Reply:
Added an explicit nil check, though it's currently a required part of the type.